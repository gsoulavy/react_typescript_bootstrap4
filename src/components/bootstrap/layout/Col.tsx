import * as React from "react";
import * as helper from "./../helpers/StatelessHelper";
import * as SuperProps  from "./../interfaces/IComponentProps";
import AlignSelf from "./enums/AlignSelf";

interface Props extends SuperProps.default {
    col?: number | boolean | string;
    sm?: number | boolean | string;
    md?: number | boolean | string;
    lg?: number | boolean | string;
    xl?: number | boolean | string;
    align?: AlignSelf;
}

const Col: React.StatelessComponent<Props> = (prop) => {
    let classResult = "";

    if (!!prop.col) {
        let subResult = "";
        if (typeof prop.col === "number") {
            subResult = `col-${prop.col}`;
        }
        if (typeof prop.col === "string" && prop.col == "auto") {
            subResult = "col-auto";
        }
        if (typeof prop.col === "boolean" && prop.col) {
            subResult = "col"
        }
        classResult = subResult;
    }

    if (!!prop.sm) {
        classResult += ` ${buildClassAttribute("sm", prop.sm)}`;
    }

    if (!!prop.md) {
        classResult += ` ${buildClassAttribute("md", prop.md)}`;
    }

    if (!!prop.lg) {
        classResult += ` ${buildClassAttribute("lg", prop.lg)}`;
    }

    if (!!prop.xl) {
        classResult += ` ${buildClassAttribute("xl", prop.xl)}`;
    }

    if (classResult.length == 0) {
        classResult = "col"
    }

    return (
        <div className={`${classResult}${helper.mergeClassName(prop.align)}${helper.mergeClassName(prop.applyClass)}`.trim()}>{prop.children}</div>
    );
};

function buildClassAttribute(size: string, value: string | boolean | number): string {
    let subResult = "";
    if (typeof value === "number") {
        subResult = `col-${size}-${value}`;
    }
    if (typeof value === "string" && value == "auto") {
        subResult = `col-${size}-auto`;
    }
    if (typeof value === "boolean" && value) {
        subResult = `col-${size}`
    }
    return subResult;

}

export default Col;