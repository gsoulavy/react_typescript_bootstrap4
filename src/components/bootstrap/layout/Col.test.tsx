import * as React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Col from "./Col";

describe("<Col />", () => {

    it("No params -> class='col'", () => {
        const component = shallow(<Col>hi</Col>);
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();
    });

    it("No params + applyClass='main' -> class='col main'", () => {
        const component = shallow(<Col applyClass="main">hi</Col>);
        expect(component.hasClass("col")).toBe(true);
        expect(component.hasClass("main")).toBe(true);
    });

    it("col='1' -> class='col-1'", () => {
        const component = shallow(<Col col={1}>hi</Col>);
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();
    });

    it("col='auto' -> class='col-auto'", () => {
        const component = shallow(<Col col="auto">hi</Col>);
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();
    });

    it("sm={true} -> class='col-sm'", () => {
        const component = shallow(<Col sm={true}>hi</Col>);
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();
    });

    it("col='true' lg={2} md='auto' -> class='col col-md-auto col-lg-2", () => {
        const component = shallow(<Col col={true} md="auto" lg={2}>hi</Col>)
        expect(component.hasClass("col")).toBe(true);
        expect(component.hasClass("col-md-auto")).toBe(true);
        expect(component.hasClass("col-lg-2")).toBe(true);
    });

});

