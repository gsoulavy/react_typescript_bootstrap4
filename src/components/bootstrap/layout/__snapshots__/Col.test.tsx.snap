// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Col /> No params -> class='col' 1`] = `
<div
  className="col"
>
  hi
</div>
`;

exports[`<Col /> align={AlignSelf.End} -> class='col align-self-end' 1`] = `
<div
  className="col align-self-end"
>
  hi
</div>
`;

exports[`<Col /> align={AlignSelf.End} applyClass='some' -> class='col align-self-end some' 1`] = `
<div
  className="col align-self-end some"
>
  hi
</div>
`;

exports[`<Col /> col='1' -> class='col-1' 1`] = `
<div
  className="col-1"
>
  hi
</div>
`;

exports[`<Col /> col='auto' -> class='col-auto' 1`] = `
<div
  className="col-auto"
>
  hi
</div>
`;

exports[`<Col /> col='true' lg={2} md='auto' -> class='col col-md-auto col-lg-2 1`] = `
<div
  className="col col-md-auto col-lg-2"
>
  hi
</div>
`;

exports[`<Col /> sm={true} -> class='col-sm' 1`] = `
<div
  className="col-sm"
>
  hi
</div>
`;
